src/stub/preprocessor/forward_declarations.rs

```diff
@@ -26,53 +26,75 @@ struct MainWrapper {
 /// Traverses through the stub commands, taking all declared variables.
 /// Leaves only one MainWrapper command in the stub.
 /// Introduces two new templates:
-/// - `forward_declaration` - similar to a read_one declares one single variable, includes all the
-///   fields in a VariableCommand (not nested under `var`)
-/// - `main_wrapper` - wraps all of the code, contains `forward_declarations` (the above resource,
-///   rendered) and `main_content`
+/// - `forward_declaration` - similar to a read_one declares one single
+///   variable, includes all the fields in a VariableCommand (not nested under
+///   `var`)
+/// - `main_wrapper` - wraps all of the code, contains `forward_declarations`
+///   (the above resource, rendered) and `main_content`
 pub fn transform(stub: &mut Stub) {
     let mut max_nested_depth = 0;
 
-    let mut forward_declarations: Vec<VariableCommand> = stub.commands.iter().filter_map(|cmd| {
+    let mut forward_declarations: Vec<VariableCommand> = stub
+        .commands
+        .iter()
+        .filter_map(|cmd| {
         let (cmd, nested_depth) = unpack_cmd(cmd, 0);
 
         if nested_depth > max_nested_depth {
             max_nested_depth = nested_depth;
         }
 
-        if let Cmd::Read(vars) = cmd {
-            Some(vars.into_iter())
+            if let Cmd::Read(var_cmds)
+            | Cmd::LoopLine {
+                variables: var_cmds, ..
+            } = cmd
+            {
+                Some(var_cmds.into_iter())
         } else {
             None
         }
-    }).flatten().collect();
+        })
+        .flatten()
+        .collect();
 
-    let mut loop_vars: Vec<VariableCommand> = ALPHABET[0..max_nested_depth].iter().filter(|loop_var|
-        forward_declarations.iter().all(|var_cmd| var_cmd.ident != loop_var.to_string())
-    ).map(|loop_var|
-        VariableCommand { 
+    let mut loop_vars: Vec<VariableCommand> = ALPHABET[0..max_nested_depth - 1]
+        .iter()
+        .filter(|loop_var| forward_declarations.iter().all(|var_cmd| var_cmd.ident != loop_var.to_string()))
+        .map(|loop_var| VariableCommand {
             ident: loop_var.to_string(), 
             var_type: VarType::Int, 
             max_length: None, 
-            input_comment: String::new() 
-        }
-    ).collect();
+            input_comment: String::new(),
+        })
+        .collect();
 
     forward_declarations.append(&mut loop_vars);
```

Comment:
If you were wondering why I didn't write this to begin with it was because of the `filter()` call having an immutable reference to `forward_declarations` while `.extend` had a mutable reference to it, causing the compiler to throw a fit.

Reply:
Changed.